src/pipelines/pipeline.py

```diff
@@ -1,27 +1,113 @@
+        is_int8 = args.dtype == torch.int8
+        if is_int8:
+            check_unused(args, {"device": torch.device("cuda")}, enforce=True)
+        torch_dtype = torch.float16 if is_int8 else args.dtype
+
+        log_rank_n("*** Creating model", logger.info)
+        with fast_init(self.device):
+            self.model = model_class._from_config(config=config, torch_dtype=torch_dtype)
+        log_rank_n("*** Moving to device", logger.info)
+        self.model.to(self.device)
+        log_rank_n("*** Initializing weights", logger.info)
+        # Initialization is ~1000x faster on GPU.
+        self.model.init_weights()
+
+        # Int8 can only be obtained by reloading a pretrained model
+        if is_int8:
+            log_rank_n("*** Saving model", logger.info)
+            self.model.save_pretrained("tmp")
+            self.model = None
+            gc.collect()
+            torch.cuda.empty_cache()
+            log_rank_n("*** Reloading model in int8", logger.info)
+            with fast_init(self.device):
+                self.model = model_class.from_pretrained(
+                    "tmp",
+                    load_in_8bit=True,
+                    device_map="auto",
+                )
+
+        self.model.eval()
+
+    def get_config(self, args) -> Tuple[Type[PreTrainedModel], PretrainedConfig]:
+        config_args = {
+            "activation_function": args.activation_function,
+            "n_head": args.n_head,
+            "n_layer": args.n_layer,
+            "bos_token_id": self.tokenizer.bos_token_id,
+            "eos_token_id": self.tokenizer.eos_token_id,
+            "vocab_size": len(self.tokenizer),
+            "use_cache": True,
+        }
+        if args.model_class.lower() == "bloom":
+            check_unused(args, {"attention_type": 1, "n_positions": None})
+            config_args["attention_softmax_in_fp32"] = True
+            config_args["hidden_size"] = args.hidden_size
+            model_class = BloomForCausalLM
+        elif args.model_class.lower() == "gpt2":
+            config_args["attention_type"] = args.attention_type
+            config_args["n_embd"] = args.hidden_size
+            config_args["n_positions"] = args.n_positions
+            config_args["print_details"] = args.print_details
+            model_class = GPT2LMHeadModel
+        else:
+            raise NotImplementedError()
+
+        return model_class, model_class.config_class(**config_args)
+
+    def __call__(self, text: List[str], **generate_kwargs) -> Tuple[List[str], Dict[str, Any]]:
+        t0 = time.perf_counter()
         input_tokens = self.tokenizer(text, return_tensors="pt", padding=True)
 
         for t in input_tokens:
             if torch.is_tensor(input_tokens[t]):
-                input_tokens[t] = input_tokens[t].to(self.input_device)
+                input_tokens[t] = input_tokens[t].to(self.device)
 
+        t1 = time.perf_counter()
         with torch.no_grad():
```

Comment:
Is this needed?
The generate method already runs in torch.no_grad context by default

Reply:
It's probably useless but a bit of extra safety can't hurt. Same with the `model.eval()` in `__init__`.

Reply:
Any reason not to use `inference_mode()`?

Reply:
I didn't know about this one, thanks for pointing it out. I will leave it for future work though.

Reply:
#12

Reply:
Sure, I'm not sure of the context here but it should be a harmless replacement ... it includes everything no_grad does.